File: Java/src/irrgarten/Player.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

import java.util.ArrayList; // import the ArrayList class

/**
 *
 * @author jorge
 */
public class Player extends LabyrinthCharacter{
    private static final int MAX_WEAPONS = 2;
    private static final int MAX_SHIELDS = 3;
    private static final int INITIAL_HEALTH = 10;
    private static final int HITS2LOSE = 3;
    
    private char number;
    private int consecutiveHits = 0;
    
    private ArrayList<Weapon> weapons;
    private ArrayList<Shield> shields;
    
    private WeaponCardDeck weaponCardDeck;
    private ShieldCardDeck shieldCardDeck;
    
    // Constructor
    public Player(char n, float i, float s){
        super("Player"+n,i,s,INITIAL_HEALTH);
        this.number = n;
        
        weapons = new ArrayList<>();
        shields = new ArrayList<>();
        
        weaponCardDeck = new WeaponCardDeck();
        shieldCardDeck = new ShieldCardDeck();
    }
    
    public Player(Player other){
        super(other);
        
        this.number = other.number;
        this.weapons= other.weapons;
        this.shields= other.shields;
        this.consecutiveHits= other.consecutiveHits;
    }
    
    // Métodos públicos
    public void resurrect(){
        weapons.clear();
        shields.clear();
        this.setHealth(INITIAL_HEALTH);
        this.resetHits();
    }
      
    public char getNumber(){
        return number;
    }
        
    public Directions move(Directions direction, ArrayList<Directions> validMoves){
        if (validMoves.size()>0 && !validMoves.contains(direction)){
            return validMoves.get(0);
        } else {
            return direction;
        }
    }
    
    @Override
    public float attack(){
        return getStrength()+this.sumWeapons();
    }
    
    @Override
    public boolean defend(float receivedAttack){
        return this.manageHit(receivedAttack);
    }
    
    public void receiveReward(){
        int wReward = Dice.weaponsReward();
        int sReward = Dice.shieldsReward();
         
        for (int i=1; i<wReward; i++){
            weapons.add(weaponCardDeck.nextCard());
        }
        
        for (int i=1; i<sReward; i++){
            shields.add(shieldCardDeck.nextCard());
        }
        
        this.setHealth(getHealth()+Dice.healthReward());
    }
    
    @Override
    public String toString(){
        String str = super.toString();
        str +="  hits: "+ String.valueOf(consecutiveHits);
        
        str+="\nweapons: ";
        for (Weapon w : weapons){
            str+= w.toString() + " ";
        }
        
        str+="\nshields: ";
        for (Shield s : shields){
            str+= s.toString() + " ";
        }
        
      //  str+= weapons.toString() + " " + shields.toString();
        str+="\n";        
        
        return str;
    }
    
    
    // Métodos privados
    private void receiveWeapon(Weapon w){
        ArrayList<Weapon> aux = new ArrayList<>(weapons);
        for(Weapon wi : weapons){
            if(wi.discard()){
                aux.remove(wi);
            }
        }
        
        weapons = aux;
        if (weapons.size()<MAX_WEAPONS) weapons.add(w);
    }

    private void receiveShield(Shield s){
        ArrayList<Shield> aux = new ArrayList<>(shields);
        for(Shield si : shields){
            if(si.discard()){
                aux.remove(si);
            }
        }
        
        shields = aux;
        if (shields.size()<MAX_SHIELDS) shields.add(s);
    }
    
    private Weapon newWeapon(){
        Weapon w = new Weapon(Dice.weaponPower(), Dice.usesLeft());
        weapons.add(w);
        return w;
    }
    
    private Shield newShield(){
        Shield s = new Shield(Dice.shieldPower(), Dice.usesLeft());
        shields.add(s);
        return s;
    }
    
    protected float sumWeapons(){
        float total=0f;
        for (Weapon w : weapons){
            total+=w.attack();
        }
        return total;
    }
    
    protected float sumShields(){
        float total=0f;
        for (Shield s : shields){
            total+=s.protect();
        }
        return total;
    }
    
    protected float defensiveEnergy(){
        return this.getIntelligence()+this.sumShields();
    }
    
    private boolean manageHit(float receivedAttack){
        if (defensiveEnergy()<receivedAttack){
            gotWounded();
            incConsecutiveHits();
        } else {
            resetHits();
        }
        
        if ((consecutiveHits == HITS2LOSE) || dead()){
            resetHits();
            return true;
        } else {
            return false;
        }
    }
    
    private void resetHits(){
        consecutiveHits=0;
    }
    
    private void incConsecutiveHits(){
        consecutiveHits++;
    }
}

File: Ruby/Player.rb
#encoding: UTF-8

require_relative 'Dice'
require_relative 'Weapon'
require_relative 'Shield'
require_relative 'LabyrinthCharacter'

module Irrgarten
	class Player < LabyrinthCharacter
		@@MAX_WEAPONS=2
		@@MAX_SHIELDS=3
		@@INITIAL_HEALTH=10
		@@HITS2LOSE=3
		
		public
		def initialize(n,i,s)
			super("Player"+n.to_s,i,s,@@INITIAL_HEALTH)
			@number=n
			@consecutive_hits = 0 
			
			@weapons = Array.new(@@MAX_WEAPONS)
          	@shields = Array.new(@@MAX_SHIELDS)
			#@weapons=[]
			#@shields=[]
		end
		
		def copy(other)
			super(other)
     		@number = other.number
          	@consecutive_hits = other.consecutive_hits
          	
          	@weapons = other.weapons
          	@shields = other.shields
		end
		
		def resurrect()
			@weapons.clear # @weapons = [] // alternativa
			@shields.clear
			@health=@@INITIAL_HEALTH
			self.reset_hits()
		end
		
		attr_reader :number,:weapons,:shields,:consecutive_hits
		
		def move(direction,valid_moves)
			contained=false
			#for dir in valid_moves
			#	if (dir==direction) contained=true
			#end
			for i in 0...valid_moves.length
  				if valid_moves[i] == direction
    					contained = true
    					break  # Salir del bucle si se encuentra el elemento
  				end
			end
			
			if (valid_moves.length>0 && !contained)
				first_element = valid_moves[0]
				return first_element
			else
				return direction
			end
		end
		
		def attack()
			@strength+self.sum_weapons
		end
		
		def defend(received_attack)
			self.manage_hit(received_attack)
		end
		
		def receive_reward()
			w_reward = Dice.weapons_reward()
			s_reward = Dice.shields_reward()
			
			w_reward.times do
				wnew = new_weapon
				receive_weapon(wnew)
			end
			
			s_reward.times do
				snew = new_shield
				receive_shield(snew)
			end
			
			@health += Dice.health_reward
		end
		
		def to_string()
			str=super+ "  hits: "+@consecutive_hits.to_s
			str+="\nweapons: "
			@weapons.each do |wi|
                	next if wi.nil?
                	str += wi.to_string
                end
                
                str +="\nshields: "
                @shields.each do |si|
                	next if si.nil?
                	str += si.to_string
                end
                
                str+="\n"
                str
		end
		
		
		private
		def receive_weapon(w)

			@weapons.reverse_each do |wi|
               	next if wi.nil?
               	if wi.discard
               		@weapons.delete(wi)
               	end
			end

			if @weapons.length < @@MAX_WEAPONS
	               #@weapons.push(w)
	               @weapons << w
			end
=begin
			@weapons = @weapons.reject { |wi| wi.discard }
			@weapons.push(w) if @weapons.size < @@MAX_WEAPONS
=end		
		end

		def receive_shield(s)

			@shields.reverse_each do |si|
               	next if si.nil?
               	if si.discard
               		@shields.delete(si)
               	end
            	end

			if @shields.length < @@MAX_SHIELDS
 	              @shields << s
	          end
=begin
			@shields = @shields.reject { |si| si.discard }
			@shields.push(s) if @shields.size < @@MAX_SHIELDS
=end		
		end
		
		def new_weapon()
			w=Weapon.new(Dice.weapon_power(), Dice.uses_left())
			@weapons << w
			w
		end
		
		def new_shield()
			s=Shield.new(Dice.shield_power(), Dice.uses_left())
			@shields << s
			s
		end
		
		def sum_weapons()
			sum=0.0
			# 0.upto(weapons.length()) do
			@weapons.each do |wi|
				next if wi.nil?
				sum+=wi.attack()
			end
			sum	
		end
		
		def sum_shields()
			sum=0.0
			@shields.each do |si|
				next if si.nil?
				sum+=si.protect()
			end
			sum
		end
		
		def defensive_energy()
			@intelligence+self.sum_shields()
		end
		
		def manage_hit(received_attack)
			defense = defensive_energy()
			lose=true
			
			if (defense<received_attack)
				self.got_wounded()
				self.inc_consecutive_hits()
			else
				self.reset_hits()
			end
			
			if ((@consecutive_hits==@@HITS2LOSE) || self.dead())
				self.reset_hits()
				lose=true
			else
				lose=false
			end
			
			lose
		end
		
		def reset_hits()
			@consecutive_hits=0
		end
			
		def inc_consecutive_hits()
			@consecutive_hits+=1
		end
		
		public_class_method :new
	end
end
